staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go

```diff
@@ -67,17 +123,128 @@ var (
+		},
+		[]string{"transformation_type", "provider_name", "key_id_hash"},
+	)
+
+	// keyIDHashLastTimestampSeconds is the last time in seconds when a keyID was used
+	// e.g. apiserver_envelope_encryption_key_id_hash_last_timestamp_seconds{key_id_hash="sha256", provider_name="providerName",transformation_type="from_storage"} 1.674865558833728e+09
+	KeyIDHashLastTimestampSeconds = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "key_id_hash_last_timestamp_seconds",
+			Help:           "The last time in seconds when a keyID was used.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"transformation_type", "provider_name", "key_id_hash"},
+	)
+
+	// keyIDHashStatusLastTimestampSeconds is the last time in seconds when a keyID was returned by the Status RPC call.
+	// e.g. apiserver_envelope_encryption_key_id_hash_status_last_timestamp_seconds{key_id_hash="sha256", provider_name="providerName"} 1.674865558833728e+09
+	KeyIDHashStatusLastTimestampSeconds = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "key_id_hash_status_last_timestamp_seconds",
+			Help:           "The last time in seconds when a keyID was returned by the Status RPC call.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"provider_name", "key_id_hash"},
+	)
 )
 
 var registerMetricsFunc sync.Once
+var hashPool *sync.Pool
 
 func RegisterMetrics() {
+	randomCacheKey := make([]byte, 32)
+	keyIDHashTotalMetricLabels = &metricLabelsMap{
+		data: make(map[string]metricLabels),
+	}
+	keyIDHashStatusLastTimestampSecondsMetricLabels = &metricLabelsMap{
+		data: make(map[string]metricLabels),
+	}
+	hashPool = &sync.Pool{
+		New: func() interface{} {
+			return hmac.New(sha256.New, randomCacheKey)
+		},
+	}
 	registerMetricsFunc.Do(func() {
-		legacyregistry.MustRegister(dekCacheFillPercent)
-		legacyregistry.MustRegister(dekCacheInterArrivals)
+		legacyregistry.MustRegister(DekCacheFillPercent)
+		legacyregistry.MustRegister(DekCacheInterArrivals)
+		legacyregistry.MustRegister(KeyIDHashTotal)
+		legacyregistry.MustRegister(KeyIDHashLastTimestampSeconds)
+		legacyregistry.MustRegister(KeyIDHashStatusLastTimestampSeconds)
 	})
 }
 
+// RecordKeyID records total count and last time in seconds when a KeyID was used for TransformFromStorage and TransformToStorage operations
+func RecordKeyID(transformationType, providerName, keyID string) {
+	h := hashPool.Get().(hash.Hash)
+	h.Reset()
+	h.Write([]byte(keyID))
+	hashPool.Put(h)
+	keyIDHash := fmt.Sprintf("sha256:%x", h.Sum(nil))
+
+	mapKey := providerName + keyIDHash
+	if !keyIDHashTotalMetricLabels.exists(mapKey) {
+		keyIDHashTotalMetricLabels.addItem(mapKey, metricLabels{
+			transformationType: transformationType,
+			providerName:       providerName,
+			keyIDHash:          keyIDHash,
+		})
+		if len(keyIDHashTotalMetricLabels.data) > maxKeyIDLabelsLimit {
```

Comment:
This is a data race, you cannot read this struct without holding the at least the read lock.

Lets make sure we have tests that run in multiple go routines so that the race detector can catch these types of mistakes.